server/src/main/java/org/opensearch/index/query/TemplateQueryBuilder.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.query;
+
+import org.apache.commons.text.StringSubstitutor;
+import org.apache.lucene.search.Query;
+import org.opensearch.common.xcontent.LoggingDeprecationHandler;
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.common.xcontent.json.JsonXContent;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.opensearch.core.xcontent.XContentParserUtils.ensureExpectedToken;
+
+/**
+ * A query builder that constructs a query based on a template and context variables.
```

Comment:
It doesn't seems like its constructing any query looking at `doToQuery()` method. I'm bit confused on its utility in core.

Reply:
Template Query will rewrite the query during doRewrite method that will direct to other queries, it won't reach to doToQuery() method in template query ideally.